lib/writers/s3.go

```diff
@@ -52,59 +52,119 @@ func NewS3Writer(logger logr.Logger, stateStoreSpec platformv1alpha1.BucketState
+		if err != nil {
+			return err
+		}
 	}
 
-	contentType := "text/x-yaml"
-	reader := bytes.NewReader(toWrite)
+	for _, item := range toWrite {
+		objectFullPath := filepath.Join(b.path, dir, item.Filepath)
+		logger := b.Log.WithValues(
+			"objectName", objectFullPath,
+		)
+		if len(toWrite) == 0 {
+			logger.Info("Empty byte[]. Nothing to write to bucket")
+			return nil
+		}
 
-	objStat, err := b.RepoClient.StatObject(ctx, b.BucketName, objectFullPath, minio.GetObjectOptions{})
-	if err != nil {
-		if minio.ToErrorResponse(err).Code == "NoSuchKey" {
-			logger.Info("Object does not exist yet")
+		ctx := context.Background()
+
+		// Check to see if we already own this bucket (which happens if you run this twice)
+		exists, errBucketExists := b.RepoClient.BucketExists(ctx, b.BucketName)
+		if errBucketExists != nil {
+			logger.Error(errBucketExists, "Could not verify bucket existence with provider")
+		} else if !exists {
+			logger.Info("Bucket provided does not exist (or the provided keys don't have permissions)")
+		}
+
+		reader := bytes.NewReader([]byte(item.Content))
+
+		objStat, err := b.RepoClient.StatObject(ctx, b.BucketName, objectFullPath, minio.GetObjectOptions{})
+		if err != nil {
+			if minio.ToErrorResponse(err).Code == "NoSuchKey" {
+				logger.Info("Object does not exist yet")
+			} else {
+				logger.Error(err, "Error fetching object")
+				return err
+			}
 		} else {
-			logger.Error(err, "Error fetching object")
+			contentMd5 := fmt.Sprintf("%x", md5.Sum([]byte(item.Content)))
+			if objStat.ETag == contentMd5 {
+				logger.Info("Content has not changed, will not re-write to bucket")
+				return nil
+			}
+		}
+
+		logger.Info("Writing object to bucket")
+		_, err = b.RepoClient.PutObject(ctx, b.BucketName, objectFullPath, reader, reader.Size(), minio.PutObjectOptions{})
+		if err != nil {
+			logger.Error(err, "Error writing object to bucket")
 			return err
 		}
-	} else {
-		contentMd5 := fmt.Sprintf("%x", md5.Sum(toWrite))
-		if objStat.ETag == contentMd5 {
-			logger.Info("Content has not changed, will not re-write to bucket")
-			return nil
+
+		// Make sure we don't delete this object, remove this object from the map
+		delete(objectsToDeleteMap, objectFullPath)
+		logger.Info("Object written to bucket")
+	}
+
+	return b.deleteObjects(ctx, objectsToDeleteMap, logger)
+}
+
+func (b *S3Writer) deleteObjects(ctx context.Context, oldObjectsToDelete map[string]minio.ObjectInfo, logger logr.Logger) error {
+	objectsCh := make(chan minio.ObjectInfo)
+	go func() {
+		defer close(objectsCh)
+		for _, objectInfo := range oldObjectsToDelete {
+			objectsCh <- objectInfo
 		}
+	}()
+
+	errorCh := b.RepoClient.RemoveObjects(ctx, b.BucketName, objectsCh, minio.RemoveObjectsOptions{})
+
+	// Print errors received from RemoveObjects API
+	var errors []error
+	for e := range errorCh {
+		logger.Error(e.Err, "Failed to remove object", "objectName", e.ObjectName)
+		errors = append(errors, e.Err)
```

Comment:
refactor: it looks like we are using this only to count the errors? maybe an `int`?